### PR TITLE
Fix env file read from include

### PR DESCRIPTION
**Pull request by jhrotko**
When reading a custom env file from include, we were not passing the read enviroenmnt variables to the config Environments, thus the read vars were being "forgotten".
Also fixed the default directory for environment variables, which should be the same as the compose.yml that is invoking the `include`

`withListeners` change forgotten in previous [PR](https://github.com/compose-spec/compose-go/pull/571)

Fixes: 
https://github.com/docker/compose/issues/11509
https://docker.atlassian.net/browse/COMP-438

### CODE REVIEW

File: loader/include.go [hunk 1]
```diff
@@ -104,17 +104,27 @@ func ApplyInclude(ctx context.Context, configDetails types.ConfigDetails, model
 			if s, err := os.Stat(f); err == nil && !s.IsDir() {
 				r.EnvFile = types.StringList{f}
 			}
+		} else {
+			envFile := []string{}
+			for _, f := range r.EnvFile {
+				f := filepath.Join(configDetails.WorkingDir, f)
```
Comment (ndeloof) [inline loader/include.go:110]: need to guard this with `filepath.IsAbs(r.Envfile)` as env_file can also be set to an absolute path, or relative to user's home `~`

File: loader/include.go [hunk 2]
```diff
@@ -104,17 +104,27 @@ func ApplyInclude(ctx context.Context, configDetails types.ConfigDetails, model
...
+				if s, err := os.Stat(f); err == nil && !s.IsDir() {
```
Comment (ndeloof) [inline loader/include.go:111]: if explicitly selected env file is a Dir, then this is an error. We only ignore when selecting the _default_ `.env` file.

File: loader/include.go [hunk 3]
```diff
@@ -104,17 +104,27 @@ func ApplyInclude(ctx context.Context, configDetails types.ConfigDetails, model
...
+				}
+			}
+			r.EnvFile = envFile
 		}
 
 		envFromFile, err := dotenv.GetEnvFromFile(configDetails.Environment, r.EnvFile)
 		if err != nil {
 			return err
 		}
+		configDetails.Environment.Merge(envFromFile)
```
Comment (ndeloof) [inline loader/include.go:122]: Why this change ?
Need to clone before merge, otherwise env_file selected to parse the included model will still be present in `configDetails` and will have impacts on next include being processed

Comment (jhrotko) [reply loader/include.go]: It took me some time to understand what is the error. Currently, the only environment variable that are resolved DURING the parse is the Interpolation (`$VAR`) and this case `environment: - VAR` will only be solved AFTER the load in https://github.com/compose-spec/compose-go/blob/4c12ed69741da3fc0acecd36bf6504687a95517e/loader/loader.go#L528 
That's why changing the `Merge` to also be applied to the higher level `configDetails` works. However, this is not correct, the environment variables of `custom.env` (from the test and the issue example) should be only within the context of the `include`. I propose to apply a `resolveEnvironment` at the end of the `loadYamlModel` to have this case into consideration, without changing the overall `project.Environment`

File: loader/loader.go [hunk 4]
```diff
@@ -787,3 +788,29 @@ func convertVolumePath(volume types.ServiceVolumeConfig) types.ServiceVolumeConf
 	volume.Source = convertedSource
 	return volume
 }
+
+func resolveEnvironment(dict map[string]any, opts *Options) {
```
Comment (ndeloof) [inline loader/loader.go:792]: I was confused reviewing this PR about this func name, which is close to `Project.ResolveEnvironment` but serve a very different scenario, but I don't have a better name for now 😅
Will create a follow-up PR if I find one

Comment (glours) [reply loader/loader.go:792]: Ok I'm not alone 😅

File: loader/loader.go [hunk 5]
```diff
@@ -787,3 +788,29 @@ func convertVolumePath(volume types.ServiceVolumeConfig) types.ServiceVolumeConf
...
+	for key, value := range dict {
+		if key == "environment" {
+			if environment, ok := value.([]any); ok {
+				serviceEnvironment := []any{}
+				for _, env := range environment {
+					env, ok := env.(string)
+					if !ok {
+						continue
+					}
+					if found, ok := opts.Interpolate.LookupValue(env); ok {
```
Comment (ndeloof) [inline loader/loader.go:802]: `Interpolate` being involved here is a bit confusing. I understand the actual goal is to access environment map, which `LookupValue` captured in options. So this basically is a hack - We should better add an explicit `Environment` (or `Variables` ?) attribute to `loader.Options` so we make it explicit we rely on the `ConfigDetails.Environment` here (and maybe in other places in future)

File: loader/loader.go [hunk 6]
```diff
@@ -787,3 +788,29 @@ func convertVolumePath(volume types.ServiceVolumeConfig) types.ServiceVolumeConf
...
+					}
+
+				}
+				dict["environment"] = serviceEnvironment
+			}
+		}
+		if v, ok := value.(map[string]any); ok {
+			resolveEnvironment(v, opts)
+		}
+	}
```
Comment (glours) [inline loader/loader.go:815]: We don't need to loop on the whole `dict` map IHMO
```suggestion
if value, ok := dict["environment"]; ok {
		if environment, ok := value.([]any); ok {
			serviceEnvironment := []any{}
			for _, env := range environment {
				env, ok := env.(string)
				if !ok {
					continue
				}
				if found, ok := opts.Interpolate.LookupValue(env); ok {
					serviceEnvironment = append(serviceEnvironment, fmt.Sprintf("%s=%s", env, found))
				} else {
					serviceEnvironment = append(serviceEnvironment, env)
				}

			}
			dict["environment"] = serviceEnvironment
		}
		if v, ok := value.(map[string]any); ok {
			resolveEnvironment(v, opts)
		}
	}
```

**Review by glours — COMMENTED** [review]
I think we should directly get the value from the map and avoid looping the whole dict

File: loader/loader.go (hunk 6 above)
Comment (ndeloof) [reply loader/loader.go:815]: can also do 
```golang
switch environment := value.(type) {
case []any:
   for _, env := range environment {   ... }
case maps[string]any:
  resolveEnvironment(v, opts)
```

**Review by ndeloof — APPROVED** [review]

**Review by glours — APPROVED** [review]
LGTM
